dnf/base.py

```diff
@@ -2159,7 +2159,6 @@ def _upgrade_internal(self, query, obsoletes, reponame, pkg_spec=None):
             query.filterm(reponame=reponame)
         query = self._merge_update_filters(query, pkg_spec=pkg_spec, upgrade=True)
         if query:
-            query = query.union(installed_query.latest())
```

Comment:
I believe that this line is required for proper handling installonly packages.

Reply:
Here is the commit - https://github.com/rpm-software-management/dnf/commit/8ea35a7e82bb8808c381f27476f023ccf7d67b20

Reply:
What about to limit `installed_query` to installed with the same name like in `query`?

`query = query.union(installed_query.latest().filterm(name=[pkg.name for pkg in query]))`

Reply:
We do not add istalled for installonly packages (for them we limit installed to latest() but to ensure the best decision by solver in combination with targeted jobs)

Reply:
I have used the solution you suggested but added one more commit for installonly pkgs, please take a look.
I have also added more tests for the situations we discussed.

Furthermore I found a problem in the previous fix so I made a follow up PR: https://github.com/rpm-software-management/libdnf/pull/1559 - this is not related to this specific bug but it concerns advisories.